### PR TITLE
Feature/datalist improve

### PR DESCRIPTION
**Pull request by LHommeDeBat**
- Algorithm-Data-List-Component uses not backend instead of hardcoded values of algorithms
- Improved style of generic data-list component
- Added sorting and searching to algorithm-list-view

### CODE REVIEW

File: src/app/components/algorithms/algorithm-list/algorithm-list.component.ts [hunk 1]
```diff
@@ -38,84 +44,140 @@ export class AlgorithmListComponent implements OnInit {
...
+      this.prepareAlgorithmdata(data);
+    });
+  }
+
+  prepareAlgorithmdata(data): void {
+    // Clear current algorithms
+    this.algorithms = [];
+
+    // Read all incoming algorithms
+    const classicAlgs = data._embedded.classicAlgorithmDtoes;
```
Comment (timniederhausen) [inline src/app/components/algorithms/algorithm-list/algorithm-list.component.ts:83]: This won't be necessary anymore.
Algorithms now end up in `_embedded.algorithms`, regardless of type.
see PlanQK/qc-atlas#99

File: generated/api/services/algorithm.service.ts [hunk 2]
```diff
@@ -61,14 +61,19 @@ export class AlgorithmService extends BaseService {
   getAlgorithms$Response(params?: {
     page?: number;
     size?: number;
-
+    sort?: string;
```
Comment (timniederhausen) [inline generated/api/services/algorithm.service.ts:64]: Don't do that. These files are generated and will be overwritten.

File: src/app/components/algorithms/algorithm-list/algorithm-list.component.ts [hunk 3]
```diff
@@ -38,84 +44,140 @@ export class AlgorithmListComponent implements OnInit {
...
 
-  searchElement(event) {
-    const searchText = event;
-    console.log(searchText);
+  searchElement(event): void {
+    this.searchParameter = event;
+    this.getAlgorithms(this.generateGetParams());
+  }
+
+  generateGetParams(): any {
```
Comment (timniederhausen) [inline src/app/components/algorithms/algorithm-list/algorithm-list.component.ts:141]: Perhaps we should move the paging / sorting stuff to the list component and add a `getParams()` there.

File: src/app/components/algorithms/algorithm-list/algorithm-list.component.ts (hunk 1 above)
Comment (LHommeDeBat) [reply src/app/components/algorithms/algorithm-list/algorithm-list.component.ts:83]: We will fix then after we merge the master, this was our workaround to test the backend

File: generated/api/services/algorithm.service.ts (hunk 2 above)
Comment (LHommeDeBat) [reply generated/api/services/algorithm.service.ts:64]: I did not want to generate everything to test the new parameter. I will do after merging the master into our branch

File: src/app/components/algorithms/algorithm-list/algorithm-list.component.ts (hunk 3 above)
Comment (LHommeDeBat) [reply src/app/components/algorithms/algorithm-list/algorithm-list.component.ts:141]: Is there a easy way to use the correct method of the correct Service in the List-Component to do the sorting/paging? Like with generics? 

Or do you mean that we just generate the params in the list and return them to the parent via event and trigger the refresh of the data in the parent? In that case we would probably have to tell the List-Component how the params look like. For example for delete the algorithm needs "algoId", publication needs "publicationId", etc.

Comment (timniederhausen) [reply src/app/components/algorithms/algorithm-list/algorithm-list.component.ts:127]: I was thinking of the second option.
Something along the lines of:
```
interface QueryParams {
page?: number;
size?: number;
sort?: string[];
// ...
}

@Output() queryChanged = new EventEmitter<QueryParams>();
```
The listener then takes the params and appends to/overwrites them as needed:
```
onQueryChanged(params: QueryParams) {
  this.algorithmService.getAlgorithms({
    ...params,
   size: 3, // we just want 3 here, regardless of what the user says
  }).subscribe(/*...*/);
}
```

Delete/Add would be a special operation (with its own EventEmitter) that just passes the object itself to the listener. The listener then needs to make the request.
Events like `pageChange`, `paginationConfigChange`, `searchElement`, `dataSorted` could be removed though.

Comment (LHommeDeBat) [reply src/app/components/algorithms/algorithm-list/algorithm-list.component.ts:127]: Got you, we will adjust that.

File: src/app/components/generics/data-list/data-list.component.ts [hunk 4]
```diff
@@ -15,84 +8,134 @@ import { SelectionModel } from '@angular/cdk/collections';
...
 export class DataListComponent implements OnInit {
   @Input() data: any[];
+  @Input() variableNames: string[];
   @Input() dataColumns: string[];
-  @Input() allowSelection: boolean;
+  @Input() allowAdd: boolean;
+  @Input() allowDelete: boolean;
   @Input() allowSearch: boolean;
-  @Input() variableNames: string[];
+  @Input() allowSort: boolean;
```
Comment (FlxB2) [inline src/app/components/generics/data-list/data-list.component.ts:16]: why don't you set default values? This shortens the HTML tag

Comment (LHommeDeBat) [reply src/app/components/generics/data-list/data-list.component.ts:16]: If you explicitly mean the boolean, it has a default = false, so if you do not want the sorting functionalty you can omit it from the HTML tag in the parent. For other optional values I will have to check

Comment (LHommeDeBat) [reply src/app/components/generics/data-list/data-list.component.ts:16]: Or we enable stuff by default and allow the parent to disable it if not needed

Comment (FlxB2) [reply src/app/components/generics/data-list/data-list.component.ts:16]: nah you are right, doesn't really make sense

**Pull request merged**